### PR TITLE
Implement BidsApp-prototype & update Dockerfile

### PR DESCRIPTION
**Pull request by khoffschlag**

### CODE REVIEW

**Comment by khoffschlag** [comment]
close #134
close #130 
close #128 
close #127 
close #126
close #125

File: pipelines/example-bids-app.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    bids_dir='../data_in/pumi-unittest/',  # Root directory of the BIDS-compliant input dataset
+    output_dir='../data_out/example-bids-app/',  # Directory where the results will be stored
+    analysis_level='participant',  # Level of the analysis that will be performed
+    participant_label=None,  # List of participant-label(s) (e.g. "001 002 003") on which pipeline is done (None -> all)
+    version=True,  # Activate/Deactivate print of PUMI-version
+    working_dir='.',  # Directory where temporary data will be stored
+).run()
+"""
+
+brain_extraction_app = BidsApp(
```
Comment (spisakt) [inline pipelines/example-bids-app.py:31]: I like this.

File: PUMI/engine.py [hunk 2]
```diff
@@ -464,3 +466,76 @@ def wrapper(name, bids_dir, subjects=None, base_dir='.', sink_dir=None, qc_dir=N
...
+
+    def __init__(self, pipeline, name, bids_dir=None, output_dir=None, analysis_level=None, participant_label=None,
+                 working_dir=None, run_args=None, description=None, **kwargs):
+
+        if description is None:
+            self.parser = argparse.ArgumentParser()
+        else:
+            self.parser = argparse.ArgumentParser(description=description)
+
+        self.parser.add_argument('--bids_dir', required=False, help='Root directory of the BIDS-compliant input dataset.')
```
Comment (spisakt) [inline PUMI/engine.py:480]: Can't we just set required to True by default?

File: pipelines/rpn_signature.py [hunk 3]
```diff
@@ -203,12 +199,9 @@ def rpn(wf, **kwargs):
...
-}
-
 
-if __name__ == '__main__':
-    print("Starting RPN-signature...")
-    rpn_wf = rpn('rpn', bids_dir=input_dir, subjects=['001'], run_args=run_args)
+rpn_app = BidsApp(
+    pipeline=rpn,
+    name='rpn',
+    bids_dir='../data_in/pumi-unittest'  # if you pass a cli argument this will be written over!
```
Comment (spisakt) [inline pipelines/rpn_signature.py:206]: we could call it default_bids_dir then?

File: pipelines/rpn_signature.py [hunk 4]
```diff
@@ -203,12 +199,9 @@ def rpn(wf, **kwargs):
...
-run_args = {
-    'plugin':'MultiProc',
-    'plugin_args':{'n_procs':4,'memory_gb':5}
-}
-
 
-if __name__ == '__main__':
-    print("Starting RPN-signature...")
-    rpn_wf = rpn('rpn', bids_dir=input_dir, subjects=['001'], run_args=run_args)
+rpn_app = BidsApp(
```
Comment (spisakt) [inline pipelines/rpn_signature.py:203]: where can we provoide runtime params like "MultiProc", N_procs, etc..?

Comment (spisakt) [reply pipelines/rpn_signature.py:203]: oh is it run_args?

Comment (spisakt) [reply pipelines/rpn_signature.py:203]: can we than create a BIDSapp where run_args can be set from the command line?

**Review by spisakt — APPROVED** [review]
Hi Kevin, nicce work, as always!
I accept it but have some comments.
We will try it out and then we can get bacck to possible enhancements.

File: PUMI/engine.py (hunk 2 above)
Comment (khoffschlag) [reply PUMI/engine.py:480]: That would be possible, but then the parameter really has to be passed via the cli. It's not possible that we set this parameter to True and then the developer sets the bids_dir in the BidsApp constructor for the end user. But it's possible that both specify a bids_dir and we select one of the inputs. If this is not a problem, we can change this.

Why this is not possible without problems: The ArgumentParser gets the inputs via sys.argv. We would have to manipulate the environment variable sys.argv at runtime for ArgumentParser not to throw an error if the user does not specify the bids_dir, which is not recommended.

File: pipelines/rpn_signature.py (hunk 4 above)
Comment (khoffschlag) [reply pipelines/rpn_signature.py:203]: Currently, it is not possible to pass run_args via the cli, only via the BidsApp constructor. But I can gladly change that.

If run_args is specified via cli as well as via the constructor, we decide to use the run_args specified by the cli, right?

**Pull request merged**